# Preliminary analysis of the Dataset

I just downloaded the dataset and want to determine the structure and the nature of variables in it

In [2]:
import pandas as pd

data = pd.read_json('dataset_airbnb-scraper_2023-06-18_04-11-45-096.json')
data.head(2)

url  \
0  https://www.airbnb.com/rooms/40751388   
1  https://www.airbnb.com/rooms/48645841   

                                                name  stars  numberOfGuests  \
0  Zen 2 bedroom apartment in the heart of El Pob...   4.96               4   
1              Private Jacuzzi Apartment in Provenza   4.64               6   

                         address                  roomType  \
0  Medellín, Antioquia, Colombia  Alojamiento entero: piso   
1  Medellín, Antioquia, Colombia  Alojamiento entero: piso   

                             location  \
0  {'lat': 6.20859, 'lng': -75.56774}   
1  {'lat': 6.20749, 'lng': -75.56672}   

                                             reviews  \
0  [{'author': {'firstName': 'Freddy', 'hasProfil...   
1  [{'author': {'firstName': 'Carlos', 'hasProfil...   

                                             pricing  \
0  {'rate': {'amount': 203, 'amountFormatted': '$...   
1  {'rate': {'amount': 179, 'amountFormatted': '$...   

                                              photos  ...  \
0  [{'caption': '', 'pictureUrl': 'https://a0.mus...  ...   
1  [{'caption': '', 'pictureUrl': 'https://a0.mus...  ...   

                                additionalHouseRules sortedReviews  \
0  - Check in 16:00 PM - Check out 11:00 AM.\n- A...            []   
1  •\tWe are Guest friendly (only one visitor per...            []   

   documentDisplayPictures  sections metadata  \
0                       []        []       {}   
1                       []        []       {}   

                         p3ImpressionId     idStr  \
0  ebcd798b-56b9-245b-9663-269436fe77ef  40751388   
1  fafbee62-1136-4ef7-55cb-49f3b0c97cf6  48645841   

  paidGrowthRemarketingListingIdsStr  \
0                                 []   
1                                 []   

                                       hostGuidebook accessibilityModule  
0                                                NaN                 NaN  
1  {'guidebookUrl': '/s/guidebooks?refinement_pat...                 NaN  

[2 rows x 85 columns]

In [3]:
data.columns

Index(['url', 'name', 'stars', 'numberOfGuests', 'address', 'roomType',
       'location', 'reviews', 'pricing', 'photos', 'primaryHost',
       'additionalHosts', 'isHostedBySuperhost', 'isAvailable', 'calendar',
       'occupancyPercentage', 'listingAmenities', 'rootAmenitySections',
       'seeAllAmenitySections', 'bathroomLabel', 'bedLabel', 'bedroomLabel',
       'hasHouseRules', 'id', 'isNewListing', 'license', 'listingRooms',
       'p3SummaryAddress', 'roomTypeCategory', 'sectionedDescription',
       'tierId', 'guestControls', 'hasSpecialOffer', 'maxNights', 'minNights',
       'priceDetails', 'hasWeWorkLocation', 'isBusinessTravelReady',
       'hometourRooms', 'hometourSections', 'descriptionLocale',
       'initialDescriptionAuthorType', 'localizedCheckInTimeWindow',
       'localizedCheckOutTime', 'city', 'country', 'countryCode',
       'hasHostGuidebook', 'hasLocalAttractions', 'neighborhoodCommunityTags',
       'state', 'paidGrowthRemarketingListingIds', 'hasCommercial

In [4]:
import numpy as np
for c in data.columns:
    if type(data[c][0]) == str:
        print('|'+c+'|str|')
    elif isinstance(data[c][0], np.floating):
        print('|'+c+'|float|')
    elif isinstance(data[c][0], np.int64):
        print('|'+c+'|int|')
    elif type(data[c][0]) == list:
        print('|'+c+'|list|')
    elif type(data[c][0]) == dict:
        print('|'+c+'|dict|')
    elif isinstance(data[c][0], np.bool_):
        print('|'+c+'|bool|')

|url|str|
|name|str|
|stars|float|
|numberOfGuests|int|
|address|str|
|roomType|str|
|location|dict|
|reviews|list|
|pricing|dict|
|photos|list|
|primaryHost|dict|
|additionalHosts|list|
|isHostedBySuperhost|bool|
|isAvailable|float|
|calendar|list|
|occupancyPercentage|float|
|listingAmenities|list|
|rootAmenitySections|list|
|seeAllAmenitySections|list|
|bathroomLabel|str|
|bedLabel|str|
|bedroomLabel|str|
|hasHouseRules|bool|
|id|int|
|isNewListing|bool|
|license|float|
|listingRooms|list|
|p3SummaryAddress|str|
|roomTypeCategory|str|
|sectionedDescription|dict|
|tierId|int|
|guestControls|dict|
|hasSpecialOffer|bool|
|maxNights|int|
|minNights|int|
|priceDetails|list|
|hasWeWorkLocation|bool|
|isBusinessTravelReady|bool|
|hometourRooms|list|
|hometourSections|list|
|descriptionLocale|str|
|initialDescriptionAuthorType|str|
|localizedCheckInTimeWindow|str|
|localizedCheckOutTime|str|
|city|str|
|country|str|
|countryCode|str|
|hasHostGuidebook|bool|
|hasLocalAttractions|bool|
|neigh

In [5]:
occup = data.occupancyPercentage[0]
print('Occupancy as calculated by Apify:', occup)

Occupancy as calculated by Apify: 15.31


In [6]:
## Making sure the occupancyPercentage variable is what I think it is

df = pd.DataFrame.from_dict(data.calendar[0])
#print( df[df.available == True].count() )
oc = df[df.available == False].count() * 100 / df.count()
assert round(occup, 2) == round( oc.available, 2), "Occupancy percentages are incosistent"
print('Occupancy calculated with available days is consistent with the one in table')

Occupancy calculated with available days is consistent with the one in table


## All Variables of full-sized JSON

|Column|Type|Description|
|------|----|-----------|
|url|str||Y|
|name|str||Y|
|stars|float||Y|
|numberOfGuests|int||Y|
|address|str|e.g., 'El Poblado, Medellín, Antioquia, Colombia'|N|
|roomType|str||Y|
|location|dict|e.g., {'lat': 6.20859, 'lng': -75.56774} |Y|
|reviews|list|metadata with reviews information|
|pricing|dict|pricing information, metadata|Y - pricing.rate.amount|
|photos|list| URL of photos|
|primaryHost|dict| info of Host, metadata|
|additionalHosts|list|additional info of Host, metadata|
|isHostedBySuperhost|bool||
|isAvailable|float|1. or nan|
|calendar|list|Every day availability of listing|
|occupancyPercentage|float||
|listingAmenities|list| amenities in the premise|
|rootAmenitySections|list| XXX I do not know|
|seeAllAmenitySections|list||
|bathroomLabel|str|Numero de baños, `'2,5 baños'`|
|bedLabel|str|e.g., `'7 camas'`|
|bedroomLabel|str|e.g., `'3 dormitorios'`|
|hasHouseRules|bool|`True` or `False`|
|id|int|7-8 digit integer|
|isNewListing|bool|
|license|float|i do not know|
|listingRooms|list| type of bed in each room, metadata|
|p3SummaryAddress|str|Seems the same as `address`|
|roomTypeCategory|str| 'entire_home', 'private_room', 'hotel_room', 'shared_room'|
|sectionedDescription|dict|Text, `'authorType', 'description', 'houseRules'...`|
|tierId|int|All Zero, 0|
|guestControls|dict|`'allowsChildren', 'allowsEvents', 'allowsPets', ...`|
|hasSpecialOffer|bool|Mostly `False`|
|maxNights|int||
|minNights|int|
|priceDetails|list|Other expenses `'Gastos de limpieza'` etc.|
|hasWeWorkLocation|bool|All `False`| 
|isBusinessTravelReady|bool|All 'False'|
|hometourRooms|list|???|
|hometourSections|list|???|
|descriptionLocale|str|`'en', 'es', 'it', 'de', ...`|
|initialDescriptionAuthorType|str|All `'human'`| 
|localizedCheckInTimeWindow|str||
|localizedCheckOutTime|str|
|city|str|`'Medellín', 'Amagá', 'El Retiro', 'Envigado', ...`|
|country|str|`'Coloimbia'`|
|countryCode|str|`'Co'`|
|hasHostGuidebook|bool|All `False`|
|hasLocalAttractions|bool|All `False`|
|neighborhoodCommunityTags|list| ????|
|state|str|`'Antioquia', 'Antioquia, Ciudad de Bolivar ', ...`|
|paidGrowthRemarketingListingIds|list| ???|
|hasCommercialHostInfo|bool||
|reviewDetailsInterface|dict|`'reviewCount'` and `'reviewSummary'`|
|reviewsOrder|str|All `'language_country'`|
|nearbyAirportDistanceDescriptions|list|???|
|renderTierId|int|All 0|
|isHotel|bool|
|showReviewTag|bool|All `False`|
|isRepresentativeInventory|bool|???|
|localizedCity|str|`'Medellín', 'Amagá', 'El Retiro', ...`|
|highlights|list|Messages from the visitors|
|highlightsImpressionId|str|something like `'ebcd798b-56b9-245b-9663-269436fe77ef'`|
|categorizedPreviewAmenities|list|
|pointOfInterests|list|Empty ???|
|chinaPointsOfInterestMatcha|dict|Empty >???|
|pageViewType|str|All `'DEFAULT'`|
|previewTags|list|Empty ???|
|seeAllHometourSections|list|Empty ???|
|enableHighlightsVoting|bool|All `False`|
|listingExpectations|list|Some details to take note, e.g., 'noise', 'escaleras'|
|localizedListingExpectations|list|Similar to previous one|
|heroModule|dict|Empty ???|
|reviewsModule|dict|Inside `'localizedOverallRating'`|
|availabilityModule|dict|Empty ???|
|additionalHouseRules|str|Long text, rules|
|sortedReviews|list| Empty ???|
|documentDisplayPictures|list| Empty ???|
|sections|list| Empty ???|
|metadata|dict|Empty ???|
|p3ImpressionId|str| like `ebcd798b-56b9-245b-9663-269436fe77ef`
|idStr|int|Like `40751388`|
|paidGrowthRemarketingListingIdsStr|list|Empty ???|

In [52]:
print( data.pricing[0]['rate']['amount'] )
print( data.listingRooms[0][0]['beds'][0]['type'] )
print( data.sectionedDescription[0]['locale'])
print( len( data.sectionedDescription[0]['description'] ) )
print( data.priceDetails[0][0]['attributeType'], data.priceDetails[0][0]['value'])
print( data.reviewDetailsInterface[0]['reviewCount'])
print( 'Balcon? ', 'balc' in data.categorizedPreviewAmenities[0][1]['amenities'])
print( len(data.additionalHouseRules[0]) )

203
queen_bed
en
1464
CLEANING_FEE 69 €
105
Balcon?  True
881


In [51]:
data.heroModule

0       {'categorizedPhotos': []}
1       {'categorizedPhotos': []}
2       {'categorizedPhotos': []}
3       {'categorizedPhotos': []}
4       {'categorizedPhotos': []}
                  ...            
1993    {'categorizedPhotos': []}
1994    {'categorizedPhotos': []}
1995    {'categorizedPhotos': []}
1996    {'categorizedPhotos': []}
1997    {'categorizedPhotos': []}
Name: heroModule, Length: 1998, dtype: object

In [160]:
df = data.copy()

df.drop(columns=['address', 'reviews', 'photos', 'additionalHosts', 'rootAmenitySections'], inplace=True)
df.drop(columns=['seeAllAmenitySections', 'license', 'p3SummaryAddress', 'tierId', 'hasSpecialOffer'], inplace=True)
df.drop(columns=['hasWeWorkLocation', 'isBusinessTravelReady', 'hometourRooms', 'hometourSections'], inplace=True)
df.drop(columns=['country', 'countryCode', 'hasHostGuidebook', 'hasLocalAttractions', 'neighborhoodCommunityTags'], inplace=True)
df.drop(columns=['state', 'paidGrowthRemarketingListingIds', 'reviewsOrder', 'nearbyAirportDistanceDescriptions'], inplace=True)
df.drop(columns=['renderTierId', 'showReviewTag', 'localizedCity', 'highlights', 'highlightsImpressionId'], inplace=True)
df.drop(columns=['pointOfInterests', 'chinaPointsOfInterestMatcha', 'pageViewType', 'previewTags'], inplace=True)
df.drop(columns=['seeAllHometourSections', 'enableHighlightsVoting', 'listingExpectations'], inplace=True)
df.drop(columns=['localizedListingExpectations', 'heroModule', 'reviewsModule', 'availabilityModule'], inplace=True)
df.drop(columns=['sortedReviews', 'documentDisplayPictures', 'sections', 'metadata', 'idStr'])
df.head(3)


url  \
0  https://www.airbnb.com/rooms/40751388   
1  https://www.airbnb.com/rooms/48645841   
2  https://www.airbnb.com/rooms/49654374   

                                                name  stars  numberOfGuests  \
0  Zen 2 bedroom apartment in the heart of El Pob...   4.96               4   
1              Private Jacuzzi Apartment in Provenza   4.64               6   
2     Finca Los Sauces Moderna y Llena De Naturaleza    NaN              16   

                    roomType                            location  \
0   Alojamiento entero: piso  {'lat': 6.20859, 'lng': -75.56774}   
1   Alojamiento entero: piso  {'lat': 6.20749, 'lng': -75.56672}   
2  Alojamiento entero: villa  {'lat': 6.02727, 'lng': -75.70012}   

                                             pricing  \
0  {'rate': {'amount': 203, 'amountFormatted': '$...   
1  {'rate': {'amount': 179, 'amountFormatted': '$...   
2  {'rate': {'amount': 421, 'amount_formatted': '...   

                                         primaryHost  isHostedBySuperhost  \
0  {'id': 270110438, 'firstName': 'Derek', 'isSup...                 True   
1  {'id': 238034915, 'firstName': 'HostinGroup', ...                False   
2  {'id': 110489568, 'firstName': 'Laura', 'isSup...                False   

   isAvailable  ...                               additionalHouseRules  \
0          1.0  ...  - Check in 16:00 PM - Check out 11:00 AM.\n- A...   
1          1.0  ...  •\tWe are Guest friendly (only one visitor per...   
2          NaN  ...                                                      

   sortedReviews documentDisplayPictures sections metadata  \
0             []                      []       []       {}   
1             []                      []       []       {}   
2             []                      []       []       {}   

                         p3ImpressionId     idStr  \
0  ebcd798b-56b9-245b-9663-269436fe77ef  40751388   
1  fafbee62-1136-4ef7-55cb-49f3b0c97cf6  48645841   
2  a2466611-90c8-e348-e4ab-f0e592ade916  49654374   

   paidGrowthRemarketingListingIdsStr  \
0                                  []   
1                                  []   
2                                  []   

                                       hostGuidebook accessibilityModule  
0                                                NaN                 NaN  
1  {'guidebookUrl': '/s/guidebooks?refinement_pat...                 NaN  
2                                                NaN                 NaN  

[3 rows x 46 columns]

In [161]:
def responsetime(x):
    if x.get('responseTime', np.nan) == 'en menos de una hora':
        return 1
    elif x.get('responseTime', np.nan) == 'en pocas horas':
        return 2
    elif x.get('responseTime', np.nan) == 'en el mismo día':
        return 3
    elif x.get('responseTime', np.nan) == 'en unos días o más':
        return 4
    elif x.get('responseTime', np.nan) == np.nan:
        return np.nan

def badges(x):
    if x['badges'] and x['badges'][0] != 'Identidad verificada':
        return int(x['badges'][0].split('\xa0')[0].replace('.',''))
    else:
        return np.nan

df['loc_lat'] = df.location.apply(lambda x: x['lat'])
df['loc_long'] = df.location.apply(lambda x: x['lng'])
df.drop(columns=['location'], inplace=True)

# the .get method allows to provide a value if the key does not exist
df['price'] = df.pricing.apply(
    lambda x: x.get('rate', {}).get('amount', np.nan) if isinstance(x, dict) and x.get('rate', {}).get('amount', 0) > 0 else np.nan)
df.drop(columns=['pricing'], inplace=True)

df['profile_pic'] = df.primaryHost.apply(lambda x: 1 if x['hasProfilePic'] == True else 0)
df['num_languages'] = df.primaryHost.apply(lambda x: len(x['languages']))
df['responseRate'] = df.primaryHost.apply(responsetime)
df['hostBadgeEvaluations'] = df.primaryHost.apply(badges)
df['hostListingCount'] = df.primaryHost.apply(lambda x: x.get('listingsCount', np.nan))
df.drop(columns=['primaryHost'], inplace=True)

df['superHost'] = df.isHostedBySuperhost.apply(lambda x: 1 if x == True else 0)
df.drop(columns=['isHostedBySuperhost'], inplace=True)

## TODO - Listing amenities
# df['calefaccion'] = 
df.drop(columns=['listingAmenities'], inplace=True)

df['numBaths'] = df.bathroomLabel.apply(lambda x: float(x.split(' ')[0].replace(',', '.')) if x and ' ' in x else np.nan)
df['numBeds'] = df.bedLabel.apply(lambda x: int(x.split(' ')[0]) if x and ' ' in x else np.nan)
df['numBedrooms'] = df.bedroomLabel.apply(lambda x: int(x.split(' ')[0]) if x and ' ' in x else np.nan)
df.drop(columns=['bathroomLabel', 'bedLabel', 'bedroomLabel'], inplace=True)



In [157]:
def banos(x):
    return x.split('')

df.bedroomLabel.apply(lambda x: int(x.split(' ')[0]) if x and ' ' in x else np.nan).unique()



array([ 2.,  3.,  8.,  4.,  7.,  1.,  5.,  6., 10., nan, 17., 18., 15.,
       11.,  9., 12., 13.])

In [113]:
def badges(x):
    if x['badges'] and x['badges'][0] != 'Identidad verificada':
        return int(x['badges'][0].split('\xa0')[0].replace('.',''))
    else:
        return 'nan'

df.primaryHost.apply(badges).unique()

array([491, 2078, 61, 30, 302, 506, 6, 15, 909, 638, 136, 1, 174, 33, 350,
       98, 76, 'nan', 223, 593, 13, 290, 755, 1229, 601, 1231, 70, 1759,
       330, 3124, 315, 73, 19, 306, 361, 12, 658, 3, 39, 401, 4, 213, 7,
       279, 1313, 309, 397, 5, 478, 183, 10, 168, 656, 155, 28, 1449, 366,
       127, 322, 162, 425, 78, 346, 233, 287, 1405, 423, 320, 193, 46, 20,
       679, 72, 1038, 5830, 116, 270, 1494, 133, 461, 137, 170, 2252, 403,
       219, 99, 236, 47, 101, 14, 231, 17, 140, 25, 38, 82, 897, 87, 27,
       781, 66, 53, 102, 1137, 2588, 35, 11, 131, 413, 363, 381, 141, 265,
       126, 339, 4907, 227, 60, 10817, 34, 1882, 498, 79, 144, 1144, 31,
       112, 1553, 1662, 4762, 357, 50, 1448, 128, 271, 32, 2136, 215, 517,
       296, 795, 2065, 293, 289, 304, 331, 199, 71, 1215, 680, 295, 1260,
       63, 159, 52, 90, 1401, 1088, 747, 1052, 216, 83, 7021, 586, 106,
       2517, 58, 4092, 134, 3563, 335, 1048, 16, 453, 67, 65, 568, 1001,
       238, 37, 771, 23, 360, 624, 3869

In [79]:
for row in df.iterrows():
    print(row[1]['pricing']['rate']['amount'])

203
179
421
178
200
275
294
188
182
192
181
214
230
189
204
245
195
195
294
202
178
180
225
187
266
275
221
237
250
238
210
186
278
185
172
549
180
430
182
511
324
518
384
336
388
243
239
228
258
340
291
414
190
350
384
305
205
275
300
386
191
428
350
408
467
382
450
360
334
176
253
312
191
250
229
320
420
194
350
354
375
189
436
170
300
467
200
212
432
185
218
271
456
309
302
344
199
292
309
376
360
358
350
375
376
375
375
404
397
316
234
206
379
192
260
279
222
179
201
220
272
189
340
177
375
350
349
397
352
328
409
320
370
419
431
374
364
388
192
328
440
425
175
360
196
241
345
341
187
319
206
191
235
179
215
185
204
179
278
175
236
194
202
236
228
175
870
1043
1358
1043
890
1199
475
961
861
800
830
834
713
481
481
474
673
541
675
481
547
499
549
548
481
554
529
580
700
600
690
602
600
594
604
751
522
495
600
475
500
631
769
697
690
500
693
700
671
721
750
482
613
500
553
699
961
979
1500
553
673
680
750
1000
1000
943
769
715
553
908
899
1417
599
586
850
851
899
600
999
492
865
705


KeyError: 'rate'